### PR TITLE
[MRV2][Spec Decode] Avoid rejection sampler OOM by chunking

### PR DESCRIPTION
**Pull request by mgoin**
## Purpose

Replacement for https://github.com/vllm-project/vllm/pull/48037 where we cap the intermediate memory required for large batch rejection sampling by using a fixed size scratch buffer and simply go through the sampling process multiple times. The loop is sync-free since chunk bounds come from `cu_num_logits_np`.

## Test Result

```
vllm serve Qwen/Qwen3-8B --spec-model RedHatAI/Qwen3-8B-speculator.dflash --spec-method dflash --spec-tokens 7 --speculative-config '{"attention_backend":"FLASHINFER"}'
```

| | peak activation | available KV cache | result |
|---|---|---|---|
| main | 3.04 GiB | 52.02 GiB | **OOM** (tried to allocate 4.12 GiB) |
| #48037 (profile it) | 7.12 GiB | 47.94 GiB | boots, loses ~4 GiB of KV cache |
| this PR | **3.04 GiB** | **52.02 GiB** | boots |

| | gsm8k num_questions=100 accuracy | mean acceptance length |
|---|---|---|
| unchunked | 0.920 | 3.06 |
| chunked | **0.920** | **3.18** |

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

### CODE REVIEW

**Comment by mgoin** [comment]
Let's go with https://github.com/vllm-project/vllm/pull/48641

File: vllm/v1/worker/gpu/spec_decode/rejection_sampler.py [hunk 1]
```diff
@@ -87,62 +113,161 @@ def _get_logprobs_tensors(
...
+                input_batch.expanded_local_pos[lo:hi],
+            )
+            chunk_logprobs = self._get_logprobs_tensors(
+                sampled,
+                num_sampled,
+                (
+                    processed_logits
+                    if self.sampler.logprobs_mode in PROCESSED_LOGPROBS_MODES
+                    else logits[lo:hi]
+                ),
```
Comment (njhill) [inline vllm/v1/worker/gpu/spec_decode/rejection_sampler.py]:
```suggestion
                processed_logits if processed else logits[lo:hi],
```
and then outside the loop, set
```python
processed = self.sampler.logprobs_mode in PROCESSED_LOGPROBS_MODES
```

File: vllm/v1/worker/gpu/spec_decode/rejection_sampler.py [hunk 2]
```diff
@@ -87,62 +113,161 @@ def _get_logprobs_tensors(
...
+                    [chunk.logprob_token_ids for chunk in logprobs_chunks]
+                ),
+                logprobs=torch.cat([chunk.logprobs for chunk in logprobs_chunks]),
+                selected_token_ranks=torch.cat(
+                    [chunk.selected_token_ranks for chunk in logprobs_chunks]
+                ),
+                cu_num_generated_tokens=(
+                    cu_num_logits_np.tolist() if expanded_logits else None
+                ),
+            )
```
Comment (njhill) [inline vllm/v1/worker/gpu/spec_decode/rejection_sampler.py]: Could move this logic into a static `LogprobsTensors.cat` method

File: vllm/v1/worker/gpu/spec_decode/rejection_sampler.py [hunk 3]
```diff
@@ -87,62 +113,161 @@ def _get_logprobs_tensors(
...
-            input_batch.expanded_idx_mapping,
-            input_batch.idx_mapping_np,
+            expanded_idx_mapping,
+            idx_mapping_np,
             pos,
             draft_sampled,
-            input_batch.expanded_local_pos,
+            expanded_local_pos,
         )
         sampled, num_sampled = rejection_sample(
```
Comment (njhill) [inline vllm/v1/worker/gpu/spec_decode/rejection_sampler.py:149]: We could consider changing to pass in sampled, num_sampled. Then those tensors could be preallocated and we just pass in chunks, to avoid the later cats.

Or might not be worth it given these are relatively small...

File: vllm/config/model.py [hunk 4]
```diff
@@ -90,6 +90,10 @@
 LogprobsMode = Literal[
     "raw_logits", "raw_logprobs", "processed_logits", "processed_logprobs"
 ]
+PROCESSED_LOGPROBS_MODES: tuple[LogprobsMode, ...] = (
```
Comment (njhill) [inline vllm/config/model.py:93]: nice

File: vllm/v1/worker/gpu/spec_decode/rejection_sampler.py [hunk 5]
```diff
@@ -87,62 +113,161 @@ def _get_logprobs_tensors(
...
+                ),
+                cu_num_generated_tokens=(
+                    cu_num_logits_np.tolist() if expanded_logits else None
+                ),
+            )
+
+        return (
+            torch.cat(sampled_chunks),
+            torch.cat(num_sampled_chunks),
+            logprobs_tensors,
```
Comment (njhill) [inline vllm/v1/worker/gpu/spec_decode/rejection_sampler.py]: nit
```suggestion
            torch.cat(sampled_chunks), torch.cat(num_sampled_chunks), logprobs_tensors
```

**Review by njhill — COMMENTED** [review]
Looks pretty good to me thanks @mgoin!

File: vllm/v1/worker/gpu/spec_decode/rejection_sampler.py [hunk 6]
```diff
@@ -87,62 +113,161 @@ def _get_logprobs_tensors(
...
+        pos: torch.Tensor,
+        max_chunk_logits: int,
+        max_num_logprobs: int,
+    ) -> tuple[torch.Tensor, torch.Tensor, LogprobsTensors | None]:
+        cu_num_logits_np = input_batch.cu_num_logits_np
+        sampled_chunks: list[torch.Tensor] = []
+        num_sampled_chunks: list[torch.Tensor] = []
+        logprobs_chunks: list[LogprobsTensors] = []
+
+        for start, end in _iter_request_chunks(cu_num_logits_np, max_chunk_logits):
```
Comment (TheEpicDolphin) [inline vllm/v1/worker/gpu/spec_decode/rejection_sampler.py:183]: With a 1GB upper bound on the target logits buffer size, it seems that chunking will be relatively rare. Would it be possible to have `_verify_in_chunks` early-return whenever no chunking is needed, and skip the `torch.cat`s?

Comment (mgoin) [reply vllm/v1/worker/gpu/spec_decode/rejection_sampler.py:183]: I originally had it separate but I wanted to prevent maintenance burden in updating two separate paths. I think the cat should be a no-op with a single tensor but I could be wrong. Will validate

File: vllm/v1/worker/gpu/spec_decode/rejection_sampler.py (hunk 1 above)
Comment (mgoin) [reply vllm/v1/worker/gpu/spec_decode/rejection_sampler.py]: good call, updated locally

File: vllm/v1/worker/gpu/spec_decode/rejection_sampler.py (hunk 2 above)
Comment (mgoin) [reply vllm/v1/worker/gpu/spec_decode/rejection_sampler.py]: good idea, moved this to LogprobsTensors.cat and reused it for prompt logprobs too

File: vllm/v1/worker/gpu/spec_decode/rejection_sampler.py (hunk 3 above)
Comment (mgoin) [reply vllm/v1/worker/gpu/spec_decode/rejection_sampler.py:149]: yeah agreed, I do not think it is worth threading buffers through the kernels for these small outputs

File: vllm/v1/worker/gpu/spec_decode/rejection_sampler.py (hunk 5 above)
Comment (mgoin) [reply vllm/v1/worker/gpu/spec_decode/rejection_sampler.py]: done locally

File: vllm/v1/worker/gpu/spec_decode/rejection_sampler.py (hunk 6 above)
Comment (mgoin) [reply vllm/v1/worker/gpu/spec_decode/rejection_sampler.py:183]: the existing single-chunk return skips the cats, so no concern

File: vllm/v1/worker/gpu/spec_decode/rejection_sampler.py [hunk 7]
```diff
@@ -87,62 +113,150 @@ def _get_logprobs_tensors(
...
+                chunk_cu_num_logits,
+                input_batch.idx_mapping[start:end],
+                input_batch.idx_mapping_np[start:end],
+                input_batch.expanded_idx_mapping[lo:hi],
+                input_batch.expanded_local_pos[lo:hi],
+            )
+            chunk_logprobs = self._get_logprobs_tensors(
+                sampled,
+                num_sampled,
+                (processed_logits if use_processed_logits else logits[lo:hi]),
```
Comment (njhill) [inline vllm/v1/worker/gpu/spec_decode/rejection_sampler.py]: redundant parens
```suggestion
                processed_logits if use_processed_logits else logits[lo:hi],
```

File: vllm/v1/worker/gpu/spec_decode/rejection_sampler.py [hunk 8]
```diff
@@ -87,62 +113,150 @@ def _get_logprobs_tensors(
...
+            del processed_logits
+            sampled_chunks.append(sampled)
+            num_sampled_chunks.append(num_sampled)
+
+        if len(sampled_chunks) == 1:
+            return (
+                sampled_chunks[0],
+                num_sampled_chunks[0],
+                logprobs_chunks[0] if logprobs_chunks else None,
+            )
```
Comment (njhill) [inline vllm/v1/worker/gpu/spec_decode/rejection_sampler.py]:
```suggestion
            logprobs_tensors = logprobs_chunks[0] if logprobs_chunks else None
            return sampled_chunks[0], num_sampled_chunks[0], logprobs_tensors
```

**Review by njhill — APPROVED** [review]
Thanks @mgoin!

just have couple more ocd style nits

**Review by TheEpicDolphin — APPROVED** [review]
LGTM
